### DISCUSSION TITLE
Using nextpnr with new FPGA which is not supported currently

### DISCUSSION THREAD

**Discussion by mustafaarslan0** [YosysHQ/nextpnr#1281] [Q&A] [answered]
Hi. I am developing a custom FPGA architecture for my thesis. I want to use nextpnr to do PnR and generate bitstream for my FPGA. But I see nextpnr-generic supports limited feature set. How can I develop nextpnr for my FPGA like others (ice40, ecp5, ..) ? Do I have to use nextpnr-generic?
Thanks.

**Comment by gatecat** [comment, marked as answer]
Depending on the size of the FPGA, I'd strongly recommend using [Viaduct](https://github.com/YosysHQ/nextpnr/blob/master/docs/viaduct.md) (below about 5-10k LUTs) or [Himbächel](https://github.com/YosysHQ/nextpnr/blob/master/docs/himbaechel.md).

Both of those get you access to the full power of the nextpnr C++ API for custom packing tasks, placement and routing validity constraints, etc; without anywhere near as much boilerplate code or maintenance overhead as starting from scratch. In both cases, there are several examples available as a starting point.

Any further questions, please do ask.

**Reply by mustafaarslan0** [reply]
Thank you for your answer. I want to know do I have to make changes in nextpnr source codes for this? Or do I just develop some database of my FPGA with Viaduct or Himbächel?

**Reply by gatecat** [reply]
In each case, at a very minimum you'll need to set up the uarch skeleton C++ code (the relevant minimal examples are a good reference here), and build the database either at runtime in C++ (Viaduct) or at build time in Python (Himbächel).

However, almost all FPGA architectures have constraints, like shared FF control sets; or clustering requirements, like carry chains. All of this will need specific C++ code as well.

**Reply by mustafaarslan0** [reply]
Thank you. I will investigate examples.
